### PR TITLE
Demo rewrite

### PR DESCRIPTION
**Pull request by Patiga**
Some relatively minor things like naming might still need to be changed
I haven't tested the demo writing yet, would be appreciated if someone else could do that.
I'm ofc open to changes :)

### CODE REVIEW

**Comment by Patiga** [comment]
I think the snapshot would be easier to work with, if it wasn't exposed as an `Iterator` or single `HashMap`, but instead as a struct with a `HashMap` for each snap object type.

```rs
struct Snapshot {
    player_input: HashMap<u16, PlayerInput>,
    projectile: HashMap<u16, Projectile>,
    laser: HashMap<u16, Laser>,
    ...
}
```
Each field would map each id to the object.
That struct could be generated using the `gamenet` python scripts, which I'm not familiar enough with.
(The fields could be suffixed with an `s` for a more intuitive plural form, but that wouldn't work all that well with all names)

**Comment by Patiga** [comment]
Rebased and with a tool to resave demos, in order to detect differences in parsing/saving.
Currently, it doesn't produce the same file, but visually the demo seemed correct. It would require further investigation, maybe Kaitai could help.

I think it is ready to merge for now. The `gamenet` and `demo` crate will be more actively used by me now, so maybe I'll need to develop this further soon.

**Comment by Patiga** [comment]
I'm not sure what to do about the 2 failing checks. They appear to be about the Rust edition 2021(?).
Should I try to lower the Rust version @heinrich5991?

**Comment by heinrich5991** [comment]
Why does it fail with nightly? 🤔

**Comment by Patiga** [comment]
A simple `cargo update` fixed the failed build. If preferred, I could probably also only run the command for `proc_macro`.

**Comment by Patiga** [comment]
After the latest change, which has nothing to do with any dependencies, `rustc-serialize` fails to build on `ubuntu-latest, nightly`???
I'll try another `cargo update -p rustc-serialize`, but I feel like I'm fixing symptoms of something I have no control over.

**Comment by Patiga** [comment]
Welp, that fixed it again

File: demo/src/format.rs [hunk 1]
```diff
@@ -98,412 +54,314 @@ impl Version {
...
     /// Tick(keyframe, tick)
-    Tick(bool, Tick),
-    Snapshot(&'a [u8]),
-    SnapshotDelta(&'a [u8]),
+    Tick {
+        tick: i32,
+        keyframe: bool,
+    },
+    Snapshot(&'a ArrayVec<[u8; MAX_SNAPSHOT_SIZE]>),
+    SnapshotDelta(&'a ArrayVec<[u8; MAX_SNAPSHOT_SIZE]>),
```
Comment (heinrich5991) [inline demo/src/format.rs:64]: Maybe rather `&'a [u8]`, I see no reason to make the type more specific.

File: snapshot/src/snap.rs [hunk 2]
```diff
@@ -227,6 +241,63 @@ impl Snap {
...
+        }
+        if let Some(last) = prev_offset {
+            if last > header.data_size {
+                return Err(Error::InvalidOffset);
+            }
+            self.sizes.push(header.data_size - last)
+        }
+
+        let mut builder = snap.recycle();
+        for size in &self.sizes {
```
Comment (heinrich5991) [inline snapshot/src/snap.rs:283]:
```suggestion
        for &size in &self.sizes {
```
gets rid of the need to dereference below.

File: doc/snapshot.md [hunk 3]
```diff
@@ -42,6 +42,30 @@ The checksum of a snapshot is calculated by summing all the `data` array
...
+    snapshot:
+        [ 4] data_size
+        [ 4] num_items
+        [*4] item_offsets
+        [  ] items
+
+Snapshots provide the full game state. In demos, they are used to provide
+additional entry points for the stream of snapshot deltas.
+
+- `data_size` is a positive signed 32-bit integer describing the amount of
```
Comment (heinrich5991) [inline doc/snapshot.md:57]: Here and below: nonnegative or positive?
(0 is nonnegative but not positive.)

File: doc/demo.md [hunk 4]
```diff
@@ -66,6 +66,8 @@ new tick, it looks like follows (sizes specified in bits):
 `is_tick` is set to `0` for normal chunks.
 
 `type` can be one of `1` (snapshot), `2` (message) or `3` (snapshot delta).
+The `chunk_data` of snapshots and snapshot deltas goes through variable-int
+compression once, and twice for messages.
```
Comment (heinrich5991) [inline doc/demo.md:70]: Does "twice" mean decoded as ints twice, and the resulting bytes are then unpacked using TW variable int encoding a third time if the message has an int?

Or does it only go through int decoding once if the message doesn't contain any integers?

File: demo/src/ddnet.rs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+use crate::reader;
+use crate::format;
+
+#[derive(Error, Debug)]
+pub enum ReadError {
+    #[error(transparent)]
+    Inner(#[from] reader::ReadError),
+    #[error("{0:?}")]
+    Snap(snapshot::snap::Error),
+    #[error("Uuid item has an incorrect size")]
```
Comment (heinrich5991) [inline demo/src/ddnet.rs:22]:
```suggestion
    #[error("UUID item has an incorrect size")]
```

File: demo/src/ddnet.rs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    UuidItemLength,
+    #[error("Uuid type id that is not registered")]
```
Comment (heinrich5991) [inline demo/src/ddnet.rs:24]:
```suggestion
    #[error("UUID type ID that is not registered")]
```

**Review by heinrich5991 — APPROVED** [review]
Looks good to me. Is this ready to merge?

I left a couple of nits, but this can be merged without fixing them. In this case, I'd go over these myself and make a PR. :)

**Comment by heinrich5991** [comment]
I didn't know about the `binrw` crate. It looks pretty nice: https://docs.rs/binrw/0.13.3/binrw/.

**Pull request merged**